In [24]:
import requests

In [25]:
product = requests.get('https://fake-store-api.mock.beeceptor.com/api/products').json()

In [26]:
# Stroing Data from Prodcut API
price = []
discount = []

for i in product:
    price.append(i['price'])
    discount.append(i['discount'])

In [27]:
print(price)
print(discount)

[599.99, 999.99, 149.99, 199.99, 499.99, 799.99, 299.99, 399.99, 599.99]
[10, 5, 15, 10, 5, 8, 12, 10, 5]


In [28]:
import numpy as np

In [29]:
# Convert list into numpy array for better calculation
price_np = np.array(price)
discount_np = np.array(discount)

In [30]:
# Calculating discount price using numpy
price_after_dis = price_np - (price_np * discount_np / 100)

In [31]:
# converting numpy array into python list
discount_price = price_after_dis.tolist()

In [32]:
Details_price = []
for i in range(len(price_np)):
    Details_price.append({
        "Price before discount": price[i],
        "Price after discount": discount_price[i],
        "On discount %": discount[i]
    })

In [33]:
for detail in Details_price:
    print(detail)

{'Price before discount': 599.99, 'Price after discount': 539.991, 'On discount %': 10}
{'Price before discount': 999.99, 'Price after discount': 949.9905, 'On discount %': 5}
{'Price before discount': 149.99, 'Price after discount': 127.4915, 'On discount %': 15}
{'Price before discount': 199.99, 'Price after discount': 179.991, 'On discount %': 10}
{'Price before discount': 499.99, 'Price after discount': 474.9905, 'On discount %': 5}
{'Price before discount': 799.99, 'Price after discount': 735.9908, 'On discount %': 8}
{'Price before discount': 299.99, 'Price after discount': 263.9912, 'On discount %': 12}
{'Price before discount': 399.99, 'Price after discount': 359.991, 'On discount %': 10}
{'Price before discount': 599.99, 'Price after discount': 569.9905, 'On discount %': 5}


# Compare the time taken by python list operation VS numpy

In [36]:
import time

# Calculate price after discount using  Python lists
start_time_list = time.time()
price_after_dis_list = []
for p, d in zip(price, discount):
    price_after_dis_list.append(p - (p * d / 100))

end_time_list = time.time()
time_list = end_time_list - start_time_list

print(f"\nTime taken for Python list operations: {time_list:.6f} seconds")


Time taken for Python list operations: 0.000133 seconds


In [46]:
start_time_np = time.time()

price_after_dis_np_time = price_np - (price_np * discount_np / 100)

end_time_np = time.time()
time_np = end_time_np - start_time_np

print(f"\nTime taken for NumPy operations calculation : {time_np:.6f} seconds")


Time taken for NumPy operations calculation : 0.000126 seconds


### Summary of Performance Comparison

*   **Time taken for Python list operations:** `0.000133` seconds
*   **Time taken for NumPy operations:** `0.000126` seconds

As observed, for this small dataset, NumPy is slightly faster than Python list operations. NumPy's efficiency significantly increases with larger datasets due to its vectorized operations.